revocation: more sophisticated revocation checking.

cpu opened the pull request:
This branch resolves the issues highlighted in https://github.com/rustls/webpki/issues/121, enforcing that certificate CRL coverage is handled properly, and unknown revocation status can be treated as an error condition when appropriate.

### revocation: policy for unknown revocation status.
This commit reworks the CRL revocation checking interface to allow specifying more policy for how CRLs should be used with a builder-based API.

In particular:

* The depth of revocation checking can be controlled, allowing users to only check the revocation status of end entity certificates.
* Handling of unknown revocation status can be controlled, allowing users to decide if unknown status should be treated as an error or  not.
* We can ensure that when revocation options are provided, there is always at least one CRL to use.

### revocation: use cert CRL DP and CRL IDP extensions.
This commit updates the CRL validation process so that when selecting a CRL we consider more than just whether the cert issuer matches the CRL issuer.

As of this commit we consider a CRL authoritative for a certificate when:

* The certificate issuer matches the CRL issuer and,
* The certificate has no CRL distribution points, and the CRL has no issuing distribution point extension.
* Or, the certificate has no CRL distribution points, but the the CRL has an issuing distribution point extension with a scope that includes the certificate.
* Or, the certificate has CRL distribution points, and the CRL has an issuing distribution point extension with a scope that includes the certificate, and at least one distribution point full name is a URI type general name that can also be found in the CRL issuing distribution point full name general name sequence.

In all other circumstances the CRL is not considered authoritative. If we can't find an authoritative CRL from the set of loaded CRLs the certificate will be considered as having unknown revocation status. Whether this is an error or not is controlled by the user provided revocation policy.

### tests: integration tests for new revocation logic.
Extends existing integration tests for new revocation logic. Also de-duplicates the owned/borrowed test generation.

### misc: clippy fixes

Addresses some `clippy` findings that bubbled up while working on this branch.

djc commented on src/crl.rs:
```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
+        ) {
+            // If the certificate has no CRL distribution points, and the CRL has no issuing distribution point,
+            // then we can consider this CRL authoritative based on the issuer matching.
+            (None, None) => true,
```
We can group these two arms: `(cert_dps, None) => cert_dps.is_none()`. That seems like a more compact representation of the original intent.

cpu replied:
Done.

cpu commented on src/end_entity.rs:
```diff
@@ -83,15 +83,15 @@ impl<'a> EndEntityCert<'a> {
         time: Time,
         eku: KeyUsage,
-        crls: &[&dyn CertRevocationList],
+        revocation: Option<RevocationOptions>,
```
I'll stage a `version = "0.102.0-alpha.0"` change and we can see whether this or #134 lands first.

cpu replied:
Done.

cpu commented on src/verify_cert.rs:
```diff
@@ -18,12 +18,38 @@ use crate::{
 };
 
+/// Allows specifying how revocation checking with CRLs should be performed.
+pub struct RevocationOptions<'a> {
```
>  Instead, use a builder pattern to construct it?

Sure, that sounds good. I'm still unclear what "style" of builder we're preferring. I've started with an implementation that's in-between the more complicated generic-over-state approach ala Rustls `Config`, but maintains the idea of being state-based by using distinct structs per state. If there's a strong preference for the "one mutable builder struct" approach that has no states at all I can rework, but it makes the error handling more unpleasant and I'm not fully convinced its the better way.

>  Ideally RevocationPolicy would have a good default.

I'm not confident what the best default is so I've gone with the most conservative: chain depth, forbidding unknown status. I figure consumers are better off relaxing a strict default than the other way around.

djc replied:
I think in this case we could probably go with the one mutable builder struct?

cpu replied:
I put together a mutable builder struct, it does seem OK for this purpose :+1:

cpu commented on src/verify_cert.rs:
```diff
@@ -18,12 +18,38 @@ use crate::{
+
+    /// A description of how revocation status checking should be applied to certificate chains
+    /// and how to handle unknown revocation status.
+    pub policy: RevocationPolicy,
```
Good catch - addressing this w/ the builder rework.

cpu commented on src/verify_cert.rs:
```diff
@@ -141,7 +172,7 @@ fn check_signatures(
     cert_chain: &Cert,
     trust_anchor: &TrustAnchor,
-    crls: &[&dyn CertRevocationList],
+    revocation: &Option<RevocationOptions>,
```
I don't think `RevocationOptions` is `Copy` because of the `&'a [&'a dyn CertRevocationList]` field. We don't want to move it so passing by ref seemed like the right thing to do.

djc replied:
IIRC references are always `Copy`.

cpu replied:
Ah you're totally right. My mistake. Deriving `Copy` and dropping the `&` to pass by value works fine.

cpu replied:
Fixed up to pass by value.

cpu commented on src/verify_cert.rs:
```diff
@@ -194,16 +225,33 @@ fn check_crls(
+        return Ok(None);
+    }
+
+    let crl = match revocation
```
> (Which begs the question: should we return Ok(None) in the (Some(_), RevocationPolicy::ChainAllowUnknown) arm?)

Not sure I understand. I think the answer is no: in this case we want to check the crl from `Some(crl)` to see if the certificate serial is present.

djc replied:
Oh yes, we only allow unknown -- sorry about that.

djc replied:
Would still suggest using a tuple match here after binding the `crl`.

cpu replied:
Done.

djc commented on src/crl.rs:
```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
+            // the CRL IDP can be considered authoritative for one of the cert DPs.
+            (Some(cert_dps), Some(crl_idp)) => {
+                IssuingDistributionPoint::from_der(untrusted::Input::from(crl_idp))
+                    .map(|crl_idp| crl_idp.can_contain(cert) && crl_idp.authoritative_for(cert_dps))
```
Yes, I just wonder (as mentioned above) if we should merge these two operations.

cpu replied:
You're right, it makes sense to merge these. I went ahead and did that. Thanks!

cpu commented on src/verify_cert.rs:
```diff
@@ -18,12 +18,38 @@ use crate::{
+/// Describes the policy to take when checking certificate revocation status.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RevocationPolicy {
+    /// Only check the end entity (leaf) certificate's revocation status. If revocation status
```
Done.

cpu commented on src/verify_cert.rs:
```diff
@@ -194,16 +225,33 @@ fn check_crls(
-    let crl = match crls
+    // If the policy only specifies checking EndEntity revocation state and we're looking at an
+    // issuer certificate, return early without considering the certificate's revocation state.
+    if let (RevocationPolicy::EndEntity, &EndEntityOrCa::Ca(_)) =
```
Fixed.

cpu commented on src/crl.rs:
```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
+    ///       the CRL issuing distribution point full name general name sequence.
+    ///
+    /// In all other circumstances the CRL is not considered authoritative.
+    fn authoritative_for(&self, cert: &Cert<'_>) -> bool {
```
I thought of a better reason to keep this as a separate operation from finding a certificate serial: we want to have the relevant CRL in hand to verify the signature from the issuer _before_ we search the CRL for a specific serial.

That said, with more thought I don't think this new code needs to be part of the trait at all. A free-standing function that takes `crl: &dyn CertRevocationList, cert: &Cert` as input is workable. Does that seem like a reasonable compromise? I've implemented that approach in-branch if you want to look it over.

cpu commented on src/crl.rs:
```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
+            // distribution point then we check the CRL IDP scope includes the type of certificate
+            // we have, in addition to matching the issuers.
+            (None, Some(crl_idp)) => {
+                IssuingDistributionPoint::from_der(untrusted::Input::from(crl_idp))
```
Switched to the `match` style.

cpu commented on src/crl.rs:
```diff
@@ -46,6 +46,60 @@ pub trait CertRevocationList: Sealed {
+            // issuing distribution point, then we decide if the CRL is authoritative by
+            // verifying the CRL IDP scope includes the type of certificate we have, and that
+            // the CRL IDP can be considered authoritative for one of the cert DPs.
+            (Some(cert_dps), Some(crl_idp)) => {
```
I think I've implemented what you were looking for here - it helped to restructure things so that checking the CRL scope for a cert and checking the IDP against the cert's DPs was done in one function as suggested in later feedback.

I'm going to resolve this conversation but if I missed the mark please unresolve and I'll take another go at it.

cpu commented on src/crl.rs:
```diff
@@ -737,6 +788,64 @@ impl<'a> IssuingDistributionPoint<'a> {
+
+        if self.only_contains_user_certs && matches!(cert.ee_or_ca, EndEntityOrCa::EndEntity) {
+            return true;
+        }
```
I reworked to use this `match` approach, but:

* rustfmt did bump the scrutinee to its own line :-( 
* clippy called out that this is better expressed as an invocation using `matches!`

I left it in the reworked form but if the two compromises above make it less appealing I can go back to the imperative style I had before.

cpu replied:
I ended up going back to the imperative `if` style. When I was adding tests I realized we also needed a `(false, false, _)` branch and the way `rustfmt` had arranged it all made it much harder to read IMO.

cpu commented on src/x509.rs:
```diff
@@ -115,6 +114,27 @@ pub(crate) struct GeneralNames<'a> {
+    ///
+    /// General names of a different type, or invalid general names are ignored.
+    pub(crate) fn uri_name_in_common_with(&mut self, other_names: &mut GeneralNames<'_>) -> bool {
+        self.any(|name| {
```
This was a good suggestion. The "trick" I was missing before was that I can re-borrow `other_names` for the nested loop.

cpu commented:
I rebased this branch and addressed the first round of feedback but I've left is a draft because it's still missing test coverage. Will land that coverage ASAP and then flip this out of draft.

cpu commented:
Rebased and added some initial test coverage. I'm going to keep this as a draft for a little bit longer, the coverage isn't 100% there yet and the Python is getting pretty messy. Closer to its final form if someone is feeling eager to review :-)

cpu commented:
I think this branch is now ready for review. It's freshly rebased & there's now sufficient test coverage.

djc commented on src/verify_cert.rs:
```diff
@@ -19,12 +19,109 @@ use crate::{
+    AllowUnknown,
+    /// Treat unknown revocation status as an error condition, yielding
+    /// [Error::UnknownRevocationStatus].
+    ForbidUnknown,
```
Should we use `Deny` here instead of `Forbid`? Maybe hoist the `Unknown` into the type name instead of repeating it in the variants? (Suggestion: `UnknownStatusPolicy`?)

cpu replied:
That sounds good. I think having shorter variant names makes sense. Updated.

djc commented on src/verify_cert.rs:
```diff
@@ -19,12 +19,109 @@ use crate::{
+#[derive(Debug, Copy, Clone)]
+/// An opaque error indicating the caller must provide at least one CRL when building a
+/// [RevocationOptions] instance.
+pub struct CrlsRequired;
```
Because it is a unit type, it is not opaque, and callers may construct it. Might want to add a private `_dummy` field to it.

cpu replied:
I changed it to `pub struct CrlsRequired(());` - I think that will have the same effect but without needing an explicitly named field like `_dummy`.

djc replied:
Yup, good call!

djc commented on src/verify_cert.rs:
```diff
@@ -151,14 +248,14 @@ fn check_signatures(
         signed_data::verify_signed_data(supported_sig_algs, spki_value, &cert.signed_data)?;
 
-        if !crls.is_empty() {
+        if let Some(ref revocation_opts) = revocation {
```
Style nit: I think the use of `ref` is not very idiomatic at this point. Instead, take a reference of the scrutinee (as in `&revocation`).

cpu replied:
Good call, thanks. Fixed up.

djc commented on src/crl.rs:
```diff
@@ -733,12 +732,88 @@ impl<'a> IssuingDistributionPoint<'a> {
             .transpose()
     }
+
+    /// Returns true if the CRL issuing distribution point has a scope that could include the cert
```
Can we add a one-line summary at the top?

cpu replied:
Done.

djc commented on src/crl.rs:
```diff
@@ -733,12 +732,88 @@ impl<'a> IssuingDistributionPoint<'a> {
+        dp_general_names: &mut DerIterator<'a, GeneralName<'a>>,
+    ) -> bool {
+        for name in idp_general_names.flatten() {
+            if let GeneralName::UniformResourceIdentifier(uri) = name {
+                for other_name in (&mut *dp_general_names).flatten() {
+                    if let GeneralName::UniformResourceIdentifier(other_uri) = other_name {
+                        if uri.as_slice_less_safe() == other_uri.as_slice_less_safe() {
+                            return true;
+                        }
+                    }
+                }
+            }
```
Style nit: linearize this?

cpu replied:
This was my attempt at linearizing after the last round :sweat_smile: I'm not sure I see what you're looking for in terms of further updates, is there a concrete suggestion I can run with?

cpu replied:
Resolved w/ the suggestion you made in the other comment thread. Thx!

djc commented on src/crl.rs:
```diff
@@ -732,12 +732,88 @@ impl<'a> IssuingDistributionPoint<'a> {
+        dp_general_names: &mut DerIterator<'a, GeneralName<'a>>,
+    ) -> bool {
+        for name in idp_general_names.flatten() {
+            if let GeneralName::UniformResourceIdentifier(uri) = name {
+                for other_name in (&mut *dp_general_names).flatten() {
+                    if let GeneralName::UniformResourceIdentifier(other_uri) = other_name {
+                        if uri.as_slice_less_safe() == other_uri.as_slice_less_safe() {
+                            return true;
+                        }
+                    }
+                }
+            }
```
Something like this?

```suggestion
            use GeneralName::UniformResourceIdentifier;
            let uri = match name {
                UniformResourceIdentifier(uri) => uri,
                _ => continue;
            };
            
            for other_name in (&mut *dp_general_names).flatten() {
                match other_name {
                    UniformResourceIdentifier(other_uri) if uri.as_slice_less_safe() == other_uri.as_slice_less_safe() => return true,
                    _ => continue,
                }
            }
```

cpu replied:
Ahh ok, that is clearer :+1: Thanks for the help. Updated.

cpu commented:
>  cpu marked this pull request as draft 2 minutes ago
Messed up my rebase and over-squashed. Fixing that up now.

cpu commented:
> Fixing that up now.

Done.

djc commented on src/lib.rs:
```diff
@@ -42,6 +42,7 @@
 #[cfg(any(test, feature = "alloc"))]
 #[cfg_attr(test, macro_use)]
 extern crate alloc;
+extern crate core;
```
Hmm, not sure this belongs in this commit?

cpu replied:
Oops, not sure how this slipped in there. Fixed.

djc commented on src/verify_cert.rs:
```diff
@@ -19,12 +19,109 @@ use crate::{
+    /// By default revocation checking will fail if the revocation status of a certificate cannot
+    /// be determined. This can be customized using the [allow_unknown_status] method.
+    pub fn new(crls: &'a [&'a dyn CertRevocationList]) -> Result<Self, CrlsRequired> {
+        match crls.len() {
```
So this probably goes against the intuition that you've built for my feedback, but I'd probably spell this as

```rust
if crls.is_empty() {
    return Err(CrlsRequired(()));
}

Ok(Self {
    ..
})
```

Why I wouldn't use the symmetric approach here: we can follow the linearization pattern here of early-returning the error case, and in this case there's no overhead of having get a usable value out of an `Option` or `Result` for the no-error path.

There also used to be a clippy lint (I think it's no longer enabled by default) that argues against using `len() == 0` in favor of using `is_empty()` on the grounds that this is more obvious. It may not have fired for this `match` setup, but I guess it still managed to imprint itself on my brain.

Whew, way too many words on silly little syntax details. Feel free to make of this what you want.

cpu commented:
> ci / Build+test (--no-default-features, stable, ubuntu-20.04) (pull_request) Failing after 2m

<details>
<summary>Looks like a network flake:</summary>

```
Caused by:
  failed to download from `[https://index.crates.io/3/s/syn`](https://index.crates.io/3/s/syn%60)

Caused by:
  [28] Timeout was reached (Operation too slow. Less than 10 bytes/sec transferred the last [30](https://github.com/rustls/webpki/actions/runs/5727388816/job/15519692580?pr=138#step:4:31) seconds)
```
</details>

Going to kick that task.

cpu commented:
> cpu force-pushed the cpu-221-enforce-coverage branch from 9ed69c4 to 3b4f7ad 

Rebased and fixed conflicts. Queuing for merge. Thanks for the reviews!
